File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
def combi(n, k):
    li = []
    result = []
    
    def recur():
        if len(li) == k:
            result.append(li[:])
            return
        
        for i in range(n):
            if not li or i > li[-1]:
                li.append(i)
                recur()
                li.pop()
    
    recur()
    return result

def main(n, maps):
    minv = float('inf')
    combiNto2 = combi(n // 2, 2)
    for teamA in combi(n, n // 2):
        teamB = [i for i in range(n) if i not in teamA]
        sumA = sum([maps[teamA[a]][teamA[b]] + maps[teamA[b]][teamA[a]] for a,b in combiNto2])
        sumB = sum([maps[teamB[a]][teamB[b]] + maps[teamB[b]][teamB[a]] for a,b in combiNto2])
        
        minv = min(minv, abs(sumA - sumB))
        
    return minv
    

import sys

input = sys.stdin.readline

n = int(input())
maps = [[int(i) for i in input().split()] for _ in range(n)]

print(main(n, maps))

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.js
const fs = require('fs');

const [_, ..._arr] = fs.readFileSync('dev/stdin').toString().trim().split("\n");
const arr = _arr.map(s => s.split(" ").map(Number))

const range = (start, end) => {
    return Array.from({length : end - start}).map((e,i) => i);
}

const getPairSum = (team) => {
    let sumv = 0
    for (let i = 0; i < team.length - 1; i++){
        for(let j = i + 1; j < team.length; j++){
            sumv += arr[team[i]][team[j]] + arr[team[j]][team[i]]
        }
    }
    
    return sumv
}

const combi = (n, k) => {
    const result = []
    const li = []
    const recur = () => {
        if (li.length == k){
            result.push([...li])
            return
        }
        
        for (let i = 0; i < n; i++ ){
            if(li.length == 0 || i > li.at(-1)){
                li.push(i)
                recur()
                li.pop()
            }
        }
    }
    
    recur()
    
    return result
}

const main = () => {
    let minv = Infinity
    const pair = combi(Math.floor(arr.length / 2), 2);
    for(let teamA of combi(arr.length, Math.floor(arr.length / 2))){
        let teamB = range(0, arr.length).filter(i => !teamA.includes(i));
        minv = Math.min(minv, Math.abs(getPairSum(teamA) - getPairSum(teamB)));
    }
    
    console.log(minv)
}

main()